File: hw/medina/u1/hw09antSimulator/model/ant_eater.py
import random

class AntEater:
    _id_counter = 1
    
    def __init__(self, position, area):
        self.id = AntEater._id_counter
        AntEater._id_counter += 1
        self.position = position
        self.area = area
        self.state = "hungry"  # hungry, eating, sleeping
        self.ants_consumed = 0
        self.eating_ticks = 0
        self.sleeping_ticks = 0
        print(f"Ant eater {self.id} created at {position}")
    
    def act(self):
        if self.state == "hungry":
            self._wander()
        elif self.state == "eating":
            self._eat()
        elif self.state == "sleeping":
            self._sleep()
    
    def _wander(self):
        current_cell = self.area.get_cell(self.position[0], self.position[1])
        
        # Check if there are ants to eat
        if current_cell.ants:
            self.state = "eating"
            self.eating_ticks = 0
            print(f"Ant eater {self.id} found ants and started eating")
        else:
            # Move randomly
            neighbors = self.area.get_neighbor_cells(self.position[0], self.position[1])
            if neighbors and random.random() < 0.5:  # 50% chance to move
                next_cell = random.choice(neighbors)
                current_cell.remove_ant_eater(self)
                self.position = (next_cell.x, next_cell.y)
                next_cell.add_ant_eater(self)
    
    def _eat(self):
        self.eating_ticks += 1
        current_cell = self.area.get_cell(self.position[0], self.position[1])
        
        if self.eating_ticks >= 10:
            if current_cell.ants:
                # Consume an ant
                ant = current_cell.ants[0]
                current_cell.remove_ant(ant)
                if ant in ant.colony.ants:
                    ant.colony.ants.remove(ant)
                
                self.ants_consumed += 1
                print(f"Ant eater {self.id} consumed ant {ant.id}. Total: {self.ants_consumed}")
                
                if self.ants_consumed >= 50:
                    self.state = "sleeping"
                    self.sleeping_ticks = 0
                    print(f"DEBUG: Ant eater {self.id} ate 50 ants and is going to sleep")
                elif current_cell.ants:
                    self.eating_ticks = 0  # Continue eating
                else:
                    self.state = "hungry"
                    print(f"Ant eater {self.id} finished eating, going back to hungry")
            else:
                self.state = "hungry"
                print(f"Ant eater {self.id} no ants left, going back to hungry")
    
    def _sleep(self):
        self.sleeping_ticks += 1
        if self.sleeping_ticks >= 60:  # Reduced for testing
            self.state = "hungry"
            self.ants_consumed = 0
            print(f" Ant eater {self.id} woke up and is hungry again")

File: hw/cevallos/u1/hw09AntSimulator/AntEater.js
class AntEater {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.state = 'HUNGRY'; 
        this.antsEaten = 0;
        this.eatingTicks = 0;
        this.sleepingTicks = 0;
    }

    update(tick, simulator) {
        const currentCell = simulator.getCell(this.x, this.y);

        switch (this.state) {
            case 'HUNGRY':
                this.behaveHungry(simulator, currentCell);
                break;
            case 'EATING':
                this.behaveEating(tick, simulator, currentCell);
                break;
            case 'SLEEPING':
                this.behaveSleeping(tick);
                break;
        }
    }

    behaveHungry(simulator, currentCell) {
        
        if (currentCell && currentCell.ants.length > 0) {
            this.state = 'EATING';
            this.eatingTicks = 0;
        } else {
            
            this.moveRandomly(simulator);
        }
    }

    behaveEating(tick, simulator, currentCell) {
        this.eatingTicks++;

        if (this.eatingTicks >= 10) {
            
            if (currentCell && currentCell.ants.length > 0) {
                const ant = currentCell.ants[0];
                currentCell.removeAnt(ant);
                this.antsEaten++;
                
                
                ant.colony.removeAnt(ant);

                
                if (this.antsEaten >= 50) {
                    this.state = 'SLEEPING';
                    this.sleepingTicks = 0;
                } 
                
                else if (currentCell.ants.length > 0) {
                    this.eatingTicks = 0; 
                } else {
                    this.state = 'HUNGRY';
                }
            } else {
                this.state = 'HUNGRY';
            }
        }
    }

    behaveSleeping(tick) {
        this.sleepingTicks++;

        if (this.sleepingTicks >= 600) {
            this.state = 'HUNGRY';
            this.antsEaten = 0;
        }
    }

    moveRandomly(simulator) {
        const neighbors = simulator.getNeighborCells(this.x, this.y);
        if (neighbors.length > 0) {
            const randomNeighbor = neighbors[Math.floor(Math.random() * neighbors.length)];
            this.moveTo(simulator, randomNeighbor.cell.x, randomNeighbor.cell.y);
        }
    }

    moveTo(simulator, newX, newY) {
        const oldCell = simulator.getCell(this.x, this.y);
        const newCell = simulator.getCell(newX, newY);
        
        if (oldCell && newCell) {
            oldCell.removeAntEater(this);
            newCell.addAntEater(this);
            this.x = newX;
            this.y = newY;
        }
    }
}

module.exports = AntEater;